src/hooks/useTokenBalances.ts

```diff
@@ -83,28 +84,39 @@ async function _getBalances(walletInfo: WalletInfo, tokens: TokenDetails[]): Pro
   const contractAddress = depositApi.getContractAddress(networkId)
   assert(contractAddress, 'No valid contract address found. Stopping.')
 
-  const balancePromises: Promise<TokenBalanceDetails | null>[] = tokens.map((token) =>
-    fetchBalancesForToken(token, userAddress, contractAddress, networkId)
-      .then((balance) => {
+  const balancePromises: Promise<TokenBalanceDetails | null>[] = tokens.map((token) => {
+    const timeoutPromise = timeout<TokenBalanceDetails | null>({
+      timeoutErrorMsg: 'Timeout fetching balances for ' + token.address,
+    })
+
+    const fetchBalancesPromise = fetchBalancesForToken(token, userAddress, contractAddress, networkId).then(
+      (balance) => {
         const cacheKey = constructCacheKey({ token, userAddress, contractAddress, networkId })
         balanceCache[cacheKey] = balance
         return balance
-      })
-      .catch((e) => {
-        console.error('[useTokenBalances] Error for', token, userAddress, contractAddress, e)
+      },
+    )
 
-        const cacheKey = constructCacheKey({ token, userAddress, contractAddress, networkId })
+    const balancePromise = Promise.race([fetchBalancesPromise, timeoutPromise])
+
+    return balancePromise.catch((e) => {
+      console.error('[useTokenBalances] Error for', token, userAddress, contractAddress, e)
 
-        const cachedValue = balanceCache[cacheKey]
-        if (cachedValue) {
-          logDebug('Using cached value for', token, userAddress, contractAddress)
-          return cachedValue
-        }
+      const cacheKey = constructCacheKey({ token, userAddress, contractAddress, networkId })
+
+      const cachedValue = balanceCache[cacheKey]
+      if (cachedValue) {
+        logDebug('Using cached value for', token, userAddress, contractAddress)
+        return cachedValue
+      }
+
+      return null
+    })
+  })
 
-        return null
-      }),
-  )
   const balances = await Promise.all(balancePromises)
+
+  // TODO: Would be better to show the errored tokens in error state
   return balances.filter(Boolean) as TokenBalanceDetails[]
```

Comment:
We don't have to cast if we use a type-guard predicate `notEmpty`, which you made precisely for these cases, didn't you?